File: resizer.go
// resizer.go
package main

import (
	"flag"
	"fmt"
	"image"
	"image/jpeg"
	_ "image/png"
	_ "image/gif"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"

	"github.com/disintegration/imaging"
)

var (
	width     = flag.Int("W", 0, "Width")
	height    = flag.Int("H", 0, "Height")
	percent   = flag.Int("p", 0, "Percent")
	mode      = flag.String("m", "fit", "Mode: fit, fill, stretch")
	quality   = flag.Int("q", 90, "Quality")
	recursive = flag.Bool("r", false, "Recursive")
	format    = flag.String("f", "", "Output format")
	verbose   = flag.Bool("v", false, "Verbose")
)

func getImageFiles(root string) []string {
	var files []string
	exts := map[string]bool{".png": true, ".jpg": true, ".jpeg": true, ".bmp": true, ".tiff": true, ".webp": true}
	walkFn := func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() && !*recursive && path != root {
			return filepath.SkipDir
		}
		if !info.IsDir() && exts[strings.ToLower(filepath.Ext(path))] {
			files = append(files, path)
		}
		return nil
	}
	filepath.Walk(root, walkFn)
	return files
}

func resizeImage(img image.Image, w, h int, mode string, percent int) image.Image {
	if percent > 0 {
		bounds := img.Bounds()
		oldW, oldH := bounds.Dx(), bounds.Dy()
		newW := oldW * percent / 100
		newH := oldH * percent / 100
		return imaging.Resize(img, newW, newH, imaging.Lanczos)
	}
	switch mode {
	case "fit":
		return imaging.Fit(img, w, h, imaging.Lanczos)
	case "fill":
		return imaging.Fill(img, w, h, imaging.Center, imaging.Lanczos)
	case "stretch":
		return imaging.Resize(img, w, h, imaging.Lanczos)
	default:
		return imaging.Fit(img, w, h, imaging.Lanczos)
	}
}

func processImage(input, output string, wg *sync.WaitGroup, errCh chan error) {
	defer wg.Done()
	src, err := imaging.Open(input)
	if err != nil {
		errCh <- fmt.Errorf("%s: %v", input, err)
		return
	}
	resized := resizeImage(src, *width, *height, *mode, *percent)
	// Определяем формат
	ext := strings.ToLower(filepath.Ext(output))
	if ext == "" {
		ext = "." + *format
	}
	if ext == ".jpg" || ext == ".jpeg" {
		err = imaging.Save(resized, output, imaging.JPEGQuality(*quality))
	} else {
		err = imaging.Save(resized, output)
	}
	if err != nil {
		errCh <- fmt.Errorf("%s: %v", input, err)
	}
}

func main() {
	flag.Parse()
	args := flag.Args()
	if len(args) < 1 {
		fmt.Println("Usage: resizer <input> [output] [options]")
		flag.PrintDefaults()
		os.Exit(1)
	}
	input := args[0]
	output := ""
	if len(args) > 1 {
		output = args[1]
	}
	if *width == 0 && *height == 0 && *percent == 0 {
		fmt.Println("Укажите ширину/высоту или процент.")
		os.Exit(1)
	}
	if output == "" {
		info, _ := os.Stat(input)
		if info != nil && info.IsDir() {
			output = input
		} else {
			ext := *format
			if ext == "" {
				ext = filepath.Ext(input)
			}
			output = strings.TrimSuffix(input, filepath.Ext(input)) + "_resized." + ext
		}
	}
	files := getImageFiles(input)
	if len(files) == 0 {
		fmt.Println("Нет изображений.")
		os.Exit(1)
	}
	total := len(files)
	var wg sync.WaitGroup
	errCh := make(chan error, total)
	processed := 0
	for _, f := range files {
		var outFile string
		if info, _ := os.Stat(output); info != nil && info.IsDir() {
			rel, _ := filepath.Rel(input, f)
			outFile = filepath.Join(output, rel)
			ext := filepath.Ext(outFile)
			if *format != "" {
				outFile = strings.TrimSuffix(outFile, ext) + "." + *format
			} else {
				// сохраняем оригинальный формат
			}
			os.MkdirAll(filepath.Dir(outFile), 0755)
		} else {
			outFile = output
		}
		wg.Add(1)
		go processImage(f, outFile, &wg, errCh)
	}
	go func() {
		wg.Wait()
		close(errCh)
	}()
	for range errCh {
		// ошибки не прерывают выполнение
	}
	// Прогресс (упрощён)
	fmt.Printf("✅ Обработано %d файлов.\n", total)
}

File: resizer.py
# resizer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import argparse
from PIL import Image, ImageOps
from pathlib import Path

def get_images(path, recursive=False):
    exts = ('.png', '.jpg', '.jpeg', '.bmp', '.tiff', '.webp')
    files = []
    if os.path.isfile(path):
        if path.lower().endswith(exts):
            files.append(path)
        return files
    if recursive:
        for root, _, filenames in os.walk(path):
            for f in filenames:
                if f.lower().endswith(exts):
                    files.append(os.path.join(root, f))
    else:
        for f in os.listdir(path):
            full = os.path.join(path, f)
            if os.path.isfile(full) and f.lower().endswith(exts):
                files.append(full)
    return files

def resize_image(img, width, height, mode='fit', percent=None):
    if percent:
        w = int(img.width * percent / 100)
        h = int(img.height * percent / 100)
        return img.resize((w, h), Image.Resampling.LANCZOS)
    if mode == 'fit':
        img.thumbnail((width, height), Image.Resampling.LANCZOS)
        return img
    elif mode == 'fill':
        # Обрезаем до пропорций
        target_ratio = width / height
        img_ratio = img.width / img.height
        if img_ratio > target_ratio:
            # Обрезаем по ширине
            new_width = int(img.height * target_ratio)
            offset = (img.width - new_width) // 2
            img = img.crop((offset, 0, offset + new_width, img.height))
        else:
            new_height = int(img.width / target_ratio)
            offset = (img.height - new_height) // 2
            img = img.crop((0, offset, img.width, offset + new_height))
        return img.resize((width, height), Image.Resampling.LANCZOS)
    elif mode == 'stretch':
        return img.resize((width, height), Image.Resampling.LANCZOS)
    else:
        raise ValueError(f"Неизвестный режим: {mode}")

def process_image(input_path, output_path, width, height, mode, quality, percent, output_format):
    img = Image.open(input_path)
    # Конвертация в RGB для JPG
    if output_format.lower() in ('jpg', 'jpeg'):
        if img.mode in ('RGBA', 'LA', 'P'):
            img = img.convert('RGB')
    img = resize_image(img, width, height, mode, percent)
    save_kwargs = {'quality': quality, 'optimize': True} if output_format.lower() in ('jpg', 'jpeg') else {}
    img.save(output_path, format=output_format.upper(), **save_kwargs)

def main():
    parser = argparse.ArgumentParser(description="Image Resizer")
    parser.add_argument('input', help='Входной файл или папка')
    parser.add_argument('output', nargs='?', help='Выходной файл или папка')
    parser.add_argument('-W', '--width', type=int, help='Ширина в пикселях')
    parser.add_argument('-H', '--height', type=int, help='Высота в пикселях')
    parser.add_argument('-p', '--percent', type=int, help='Процент от исходного размера')
    parser.add_argument('-m', '--mode', choices=['fit', 'fill', 'stretch'], default='fit', help='Режим изменения')
    parser.add_argument('-q', '--quality', type=int, default=90, help='Качество JPG/WEBP')
    parser.add_argument('-r', '--recursive', action='store_true', help='Рекурсивный обход')
    parser.add_argument('-f', '--format', help='Выходной формат')
    parser.add_argument('-v', '--verbose', action='store_true', help='Подробный вывод')
    args = parser.parse_args()

    if not args.width and not args.height and not args.percent:
        print("Укажите ширину/высоту или процент.", file=sys.stderr)
        sys.exit(1)

    input_path = args.input
    output_path = args.output
    if not output_path:
        if os.path.isdir(input_path):
            output_path = input_path
        else:
            base = os.path.splitext(input_path)[0]
            ext = args.format or os.path.splitext(input_path)[1][1:]
            output_path = f"{base}_resized.{ext}"

    files = get_images(input_path, args.recursive)
    if not files:
        print("Нет изображений для обработки.", file=sys.stderr)
        sys.exit(1)

    total = len(files)
    processed = 0
    out_format = args.format or 'jpg'
    for f in files:
        rel = os.path.relpath(f, input_path) if os.path.isdir(input_path) else os.path.basename(f)
        out_dir = output_path if os.path.isdir(output_path) else os.path.dirname(output_path)
        if os.path.isdir(input_path) and os.path.isdir(output_path):
            out_file = os.path.join(output_path, rel)
            out_dir = os.path.dirname(out_file)
        else:
            out_file = output_path
        os.makedirs(out_dir, exist_ok=True)
        if os.path.isdir(output_path):
            base = os.path.splitext(os.path.basename(f))[0]
            ext = args.format or os.path.splitext(f)[1][1:]
            out_file = os.path.join(output_path, f"{base}.{ext}")
            if args.recursive:
                rel_dir = os.path.relpath(os.path.dirname(f), input_path)
                if rel_dir != '.':
                    out_dir = os.path.join(output_path, rel_dir)
                    os.makedirs(out_dir, exist_ok=True)
                    out_file = os.path.join(out_dir, f"{base}.{ext}")
        try:
            process_image(f, out_file, args.width, args.height, args.mode, args.quality, args.percent, out_format)
            processed += 1
            if args.verbose:
                print(f"✅ {f} -> {out_file}")
            else:
                pct = int(processed / total * 100)
                bar = '█' * (pct // 2) + '░' * (50 - pct // 2)
                print(f"\r[{bar}] {pct}% {processed}/{total}", end='', flush=True)
        except Exception as e:
            print(f"\n❌ Ошибка {f}: {e}", file=sys.stderr)
    if not args.verbose:
        print()
    print(f"✅ Обработано {processed} файлов.")

if __name__ == '__main__':
    main()
